src/compiler/checker.ts

```diff
@@ -9779,14 +9790,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     if (p.flags & SymbolFlags.Accessor && useAccessors) {
                         const result: AccessorDeclaration[] = [];
                         if (p.flags & SymbolFlags.SetAccessor) {
+                            const setter = p.declarations && forEach(p.declarations, d => {
+                                if (d.kind === SyntaxKind.SetAccessor) {
+                                    return d as SetAccessorDeclaration;
+                                }
+                                if (isCallExpression(d) && isBindableObjectDefinePropertyCall(d)) {
+                                    return forEach(d.arguments[2].properties, propDecl => {
+                                        const id = getNameOfDeclaration(propDecl);
+                                        if (!!id && isIdentifier(id) && idText(id) === "set") {
+                                            return propDecl;
+                                        }
+                                    });
+                                }
+                            });
+
+                            Debug.assert(setter && isFunctionLikeDeclaration(setter));
+                            const paramSymbol: Symbol | undefined = getSignatureFromDeclaration(setter).parameters[0];

```

Comment:
This annotation isn't enforced, it acts as a documentation hint~. The problem is that indexed access here won't return `| undefined` and TypeScript prefers the expression type here over the declared type (it's a known design decision). In practice, this array might be empty and I handle it in the implementation to make the truthiness check less weird I decided to annotate this here.